Use an ingress controller on the control-plane network for UI and API (through UI) access

NicolasT opened the pull request:
Cfr. https://github.com/scality/metalk8s/issues/1800#issuecomment-536082408

Still a bit of work to be done w.r.t. inter-service TLS traffic, and eventually turning this into a single *Ingress* object, but this should be a good start.

TODO:
- [x] Docs
- [ ] Consider using the host-network for the ingress controller, iff we can bind the resulting `nginx` to only the control-plane IP of a node, to recover source/client IPs (which doesn't work with `externalIPs` in a *Service*)

NicolasT committed:
salt: fix (once again) `Service` handling in `metalk8s_kubernetes`

This time, the `externalIPs` field was not properly propagated.

NicolasT committed:
salt: deploy `nginx-ingress-control-plane` addon

This provisions an ingress controller for the `nginx-control-plane`
class, which will run on all `master` nodes, and be exposed on port 8443
off the control-plane IP address of the `bootstrap` node. We can add
addresses of other `master` nodes later.

The `chart.sls` file was generated using

```
./charts/render.py nginx-ingress-control-plane metalk8s-ingress charts/nginx-ingress-control-plane.yaml charts/nginx-ingress/ > salt/metalk8s/addons/nginx-ingress-control-plane/deployed/chart.sls
```

Fixes: #1796
See: https://github.com/scality/metalk8s/issues/1796
See: https://github.com/scality/metalk8s/issues/1800

NicolasT committed:
salt: fix `externalName` support in `Service` deployment

Can't wait for https://github.com/scality/metalk8s/pull/1713 to land.

See: 8fa9b38ef1ee6da8d522adf625c941cb8a2a25a5

NicolasT committed:
salt: add *Ingress* support to renderer/state/module

NicolasT committed:
salt, ui: deploy UI using *Ingress*es

The Kubernetes API, SaltAPI, Prometheus and the actual UI are now
exposed using the `nginx-control-plane` *Ingress* controller, and as
such accessible through the control-plane network IP of the 'bootstrap'
node, port 8443.

Note: this also updates Cypress to a (for now) unreleased version,
because the current released version doesn't support self-signed TLS
certificates when using an IP-based host as a test target (which is the
case in our setup). This was reported upstream in
https://github.com/cypress-io/cypress/issues/771 and fixed in
https://github.com/cypress-io/cypress/pull/4947. The information as to
how to install an unreleased version of Cypress I got from
https://github.com/cypress-io/cypress/issues/4525. We also ensure all
shared libraries this version of Cypress uses are installed in the test
environment.

Fixes: #1602
See: https://github.com/scality/metalk8s/issues/1602
Fixes: #1797
See: https://github.com/scality/metalk8s/issues/1797
Fixes: #1799
See: https://github.com/scality/metalk8s/issues/1799
Fixes: #1800
See: https://github.com/scality/metalk8s/issues/1800
See: https://github.com/cypress-io/cypress/pull/4947
See: https://github.com/cypress-io/cypress/issues/771
See: https://github.com/cypress-io/cypress/issues/4525

NicolasT committed:
salt: no longer deploy Prometheus `NodePort` *Service*

bert-e commented:
# Hello nicolast,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @sayf-eddine-scality

* @thomasdanan

* @carlito-scality

* @teddyandrieux

* @nootal

* @lucieleonard

* @gdemonet

* @jbertran

* @nicolast

* @monpote

* @tcarmet

* @slaperche-scality

NicolasT committed:
ui: remove certificate acceptance links from login page

We now host the UI and all APIs it wants to access behind a single
*Ingress* object, so there's only a single TLS certificate to be
accepted.

See: https://github.com/scality/metalk8s/pull/1802

NicolasT committed:
salt: support `extensions/v1beta1/Ingress` in K8s renderer

NicolasT committed:
monitoring, salt: deploy an *Ingress* for Grafana

From now on, Grafana is accessible as `/grafana` on the
`nginx-control-plane` ingress controller.

Fixes: #1798
See: https://github.com/scality/metalk8s/issues/1798

NicolasT committed:
monitoring: disable Grafana phone-home

NicolasT commented:
This is a bit stuck on the Cypress tests, because we now expose the UI on HTTPS using an IP-address and a self-signed certificate.

I believe this could be fixed by https://github.com/cypress-io/cypress/pull/4947 (cfr. https://github.com/cypress-io/cypress/issues/771), but there's no Cypress release including this change, and I seem to be unable to install an unreleased version for testing :roll_eyes:

NicolasT commented:
Jay, found a work-around :tada:

sayf-eddine-scality commented on docs/quickstart/services.rst:
```diff
@@ -64,23 +47,8 @@ This page displays two monitoring indicators:
-Then open your web browser and navigate to ``http://localhost:3000``
-
+Grafana is available on the same host as the MetalK8s UI, under ``/grafana``.
+Log in with the default credentials: ``grafana`` / ``prom-operator``.
```
Still having issue to login the default creds are not working

NicolasT replied:
That's because I made a mistake in the docs ;-P Will fix.

NicolasT replied:
I changed the default one to `admin`/`admin`. Should get OIDC up soon!

sayf-eddine-scality commented on charts/nginx-ingress-control-plane.yaml:
```diff
@@ -0,0 +1,42 @@
+  image:
+    repository: '{%- endraw -%}{{ build_image_name("nginx-ingress-controller", False) }}{%- raw -%}'
+
+  defaultBackendService: 'metalk8s-ingress/nginx-ingress-default-backend'
```
we may need a good 404 page with the same theme as the UI

gdemonet replied:
Indeed, could even redirect to MetalK8s UI 404 page.

NicolasT replied:
Good idea, but I think that's out of scope for this PR.

sayf-eddine-scality reviewed (changes requested):
two minor comments the rest is LGTM

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @sayf-eddine-scality

* @thomasdanan

* @carlito-scality

* @teddyandrieux

* @nootal

* @lucieleonard

* @gdemonet

* @jbertran

* @nicolast

* @monpote

* @tcarmet

* @slaperche-scality





The following reviewers are expecting changes from the author, or must review again:

* @sayf-eddine-scality

NicolasT committed:
monitoring, docs: change Grafana password to default one

See: https://github.com/scality/metalk8s/pull/1802#discussion_r329952481

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @sayf-eddine-scality

* @thomasdanan

* @carlito-scality

* @teddyandrieux

* @nootal

* @lucieleonard

* @gdemonet

* @jbertran

* @nicolast

* @monpote

* @tcarmet

* @slaperche-scality

NicolasT commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye nicolast.
